rosidl_generator_cpp/rosidl_generator_cpp/__init__.py

```diff
@@ -176,7 +176,108 @@ def primitive_value_to_cpp(type_, value):
     assert False, "unknown primitive type '%s'" % type_.type
 
 
+def default_cpp_value_from_type(type_):
+    if type_ == 'string':
+        return ''
+    elif type_ in ['float32', 'float64']:
+        return '0.0'
+    elif type_ == 'bool':
+        return 'false'
+    return '0'
+
+
 def escape_string(s):
     s = s.replace('\\', '\\\\')
     s = s.replace('"', '\\"')
     return s
+
+
+def create_init_alloc_and_member_lists(spec):
+    # A Member object represents the information we need to know to initialize
+    # a single member of the class.
+    class Member:
+        def __init__(self, name):
+            self.name = name
+            self.default_value = None
+            self.zero_value = None
+            self.zero_need_array_override = False
+            self.type = None
+            self.num_prealloc = 0
+
+    # A CommonMemberSet is a set of adjacent members that share the same set of
+    # initialization semantics.  Here, initialization semantics mean that all
+    # members of the set have a default value (or do not have a default value),
+    # and all members of the set have a zero value (or do not have a zero
+    # value).
+    class CommonMemberSet:
+        def __init__(self):
+            self.members = []
+
+        def add_member(self, memb):
+            if not self.members:
+                self.members.append(memb)
+                return True
+            new_def_none = memb.default_value is None
+            last_def_none = self.members[-1].default_value is None
+            new_zero_none = memb.zero_value is None
+            last_zero_none = self.members[-1].zero_value is None
+            if (new_def_none == last_def_none) and (new_zero_none == last_zero_none):
+                self.members.append(memb)
+                return True
+            return False
```

Comment:
This code seems overly complicated to me. I would recommend something like this:

```
class Member:
    def same_default_and_zero_value(self, other):
        return self.default_value == other.default_value and self.zero_value == other.zero_value
```

This makes the code much easier to read:

```
class CommonMemberSet:
    def add_member(self, memb):
        if not self.members or self.members[-1].same_default_and_zero_value(memb):
            self.members.append(memb)
            return True
        return False
```

Reply:
Yeah, that is a much better implementation.  I've implemented it this way instead.